pkg/k8s/watchers/pod.go

```diff
@@ -271,6 +281,85 @@ func realizePodAnnotationUpdate(podEP *endpoint.Endpoint) {
 	}
 }
 
+// updateCiliumEndpointLabels runs a controller associated with the endpoint that updates
+// the Labels in CiliumEndpoint object by mirroring those of the associated Pod.
+func updateCiliumEndpointLabels(ep *endpoint.Endpoint, labels map[string]string) {
+	var (
+		controllerName = fmt.Sprintf("sync-pod-labels-with-cilium-endpoint (%v)", ep.GetID())
+		scopedLog      = log.WithField("controller", controllerName)
+	)
+
+	// The controller is executed only once and is associated with the underlying endpoint object.
+	// This is to make sure that the controller is also deleted once the endpoint is gone.
+	ep.UpdateController(controllerName,
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) (err error) {
+				capabilities := k8sversion.Capabilities()
+				pod := ep.GetPod()
+				ciliumClient := k8s.CiliumClient().CiliumV2()
+
+				switch {
+				case capabilities.Patch:
+					replaceLabels := []k8s.JSONPatch{
+						{
+							OP:    "replace",
+							Path:  "/metadata/labels",
+							Value: labels,
+						},
+					}
+
+					labelsPatch, err := json.Marshal(replaceLabels)
+					if err != nil {
+						scopedLog.WithError(err).Debug("Error marshalling Pod labels")
+						return err
+					}
+
+					_, err = ciliumClient.CiliumEndpoints(pod.GetNamespace()).Patch(
+						ctx, pod.GetName(),
+						types.JSONPatchType,
+						labelsPatch,
+						meta_v1.PatchOptions{})
+					if err != nil {
+						scopedLog.WithError(err).Debug("Error while updating CiliumEndpoint object with new Pod labels")
+						return err
+					}
+
+				default:
+					status := v2.EndpointStatus{}
+					ep.GetCiliumEndpointStatus(option.Config).DeepCopyInto(&status)
```

Comment:
nit: this can be `status := ep.GetCiliumEndpointStatus(option.Config).DeepCopy()` if the PR is merged after https://github.com/cilium/cilium/pull/12680. I'm fine leave it as is.